dist-build/android-build.sh

```diff
@@ -18,18 +18,23 @@ if [ ! -f ./configure ]; then
   exit 1
 fi
 
-if [ -z "$TARGET_ARCH" ] || [ -z "$ARCH" ] || [ -z "$HOST_COMPILER" ]; then
+if [ "x$HOST_COMPILER" = 'x' ]; then
   echo "You shouldn't use android-build.sh directly, use android-[arch].sh instead" >&2
   exit 1
 fi
 
-export PREFIX="$(pwd)/libsodium-android-${TARGET_ARCH}"
-export TOOLCHAIN_OS_DIR="$(uname | tr '[:upper:]' '[:lower:]')-x86_64/"
+if [ -z ${PREFIX+x} ]; then export PREFIX="$(pwd)/libsodium-android"; echo "setting default PREFIX=${PREFIX}"; fi
+if [ -z ${LIB_PATH_PREFIX+x} ]; then echo ""; else export LIB_PATH="${LIB_PATH_PREFIX}/${TARGET_ARCH_DIR}"; fi
+if [ -z ${LIB_PATH+x} ]; then export LIB_PATH="${PREFIX}/lib/${TARGET_ARCH_DIR}"; fi
+if [ -z ${INCLUDE_PATH+x} ]; then export INCLUDE_PATH="${PREFIX}/include"; fi
```

Comment:
Does `cmake` really have an issue with header files being in `${PREFIX}/${ARCH}/include` and libs being in `${PREFIX}/${ARCH}/lib`?

Having different prefixes makes the layout different from system libraries. It also assumes that header files are exactly the same for all targets, which may not always be the case.

Reply:
This is related to me thinking matching something similar to /usr/lib/x86_64-linux-gnu was your desired approach to simplify cross compilation. It definitely can work I think it requires using cmake variables instead of just specifying a root folder.

Reply:
I did some experimenting. As far as I know you have to specify include directories no matter how they are organized cmake won't search for folders named include or deal with architecture variables so whether it's ${PREFIX}/${ARCH}/include or or ${PREFIX}/include/${ARCH}/ it doesn't make a difference as far as I can see. For libraries find_library will search for architectures inside lib folders but not the other way around. This is easy to work around you just have to specify a different root path to search based on the same variable you need for includes anyway list(APPEND CMAKE_FIND_ROOT_PATH "${PREFIX}/${CMAKE_LIBRARY_ARCHITECTURE}")

Thinking about it a bit more and doing some more research I think I still prefer the lib/${ARCH} approach. It does match the way linux handles libraries in the /usr/lib folder as well as the way platform specific includes are handled in /usr/include. The ndk compiler for aarch64 has a default search path for includes:  $ANDROID_NDK_HOME/toolchains/llvm/prebuilt/linux-x86_64/bin/./../sysroot/usr/include/aarch64-linux-android 
it actually doesn't have one for the sysroot/usr/local/include. Researching that a bit more I now realize my concept of /usr/local/lib has been wrong my whole life. I've always put anything I compiled myself in /usr/local/ but even if I do so I should have been putting libraries in /usr/lib. So anyway if we do ${PREFIX}/lib/${ARCH} then someone could set the prefix to their ndk sysroot and then the appropriate architecture's ndk compiler will just find the includes and the libs with no need for any special handling by CMAKE or any other build tool. That option isn't available if we use ${ARCH}/lib .

thinking a bit more about your comment about matching other architectures I was thinking about the host architecture not other scripts in dist-build. they all store their files in libsodium-${arch} folders I don't think this approach really deviates as that folder is just the default $PREFIX the script as it currently exists will put all the files in libsodium-android and then have the lib/include $ARCH stuff within that folder. I don't think any of the other dist-build scripts support more than one architecture. For the automated build infrastructure this would mean the resulting files always have exactly the same names as the old scripts used to include the android version though I discuss that in the para below. It sort of touches on a different comment you made about android versioning and where the version should be set but I discuss it here as it's relevant to the preferred directory layout as well.

The only remaining weirdness for Android is the different android versions. The ndk compilers and cmake don't take these into account when looking for folders. I sort of expect that most people just compile all their ndk libs for the lowest common denominator and rely on forwards compatibility. If someone wants to have different versions of libraries that support different android versions they'll need to use some sort of build system to manage that no matter what AFAIK. I misread the original scripts I thought it was falling back to a compat version if there was a failure but it actually originally compiled a specified version (based on script variables) and then it would compile a compat version as well unless the compat version and specified version were the same. When I updated the scripts now it will actually compile the compat version and replace the version that was asked for. There aren't any useful architecture based names that include android version (ie ones that clang or cmake will find automatically) so the best way to deal with this would be to just change the prefix by appending the android version. This would of course require special handling if a custom prefix was specified. We don't want "sysroot-android21" in an ndk if anyone actually does want to use their sysroot so if a PREFIX is specified we shouldn't do the compat check... but if the compat version is the default then that means the code would never be called... I don't think this compat fallback actually makes any sense as the old scripts had to be edited to change the target version anyway so why not just have the compat version the default and if the user wants to specify something different we should trust them and not compile a second compat version.